Sources/SymbolKit/SymbolGraph/SymbolGraph+Overloads.swift

```diff
@@ -0,0 +1,126 @@
+    ///
+    /// The symbol that appears first in this sorting is then cloned to create an "overload group".
+    /// This symbol will have a unique identifier based on the original symbol, but suffixed with
+    /// ``Symbol/overloadGroupIdentifierSuffix``. New ``Relationship/Kind/overloadOf``
+    /// relationships are created between the colliding symbols and the new overload group symbol.
+    /// In addition, any existing relationships the original symbol had are also cloned for the
+    /// overload group.
+    public mutating func createOverloadGroupSymbols() {
+        struct OverloadKey: Hashable {
+            let path: [String]
+            let kind: SymbolGraph.Symbol.KindIdentifier
+        }
+
+        var symbolsByPath = [OverloadKey: [SymbolGraph.Symbol]]()
+
+        for symbol in self.symbols.values {
+            if symbol.kind.identifier.isOverloadableKind {
+                symbolsByPath[.init(path: symbol.pathComponents, kind: symbol.kind.identifier), default: []].append(symbol)
+            }
+        }
+
+        var newRelationships = [Relationship]()
+
+        for overloadSymbols in symbolsByPath.values where overloadSymbols.count > 1 {
+            let sortedOverloads: [Symbol]
+            if overloadSymbols.allSatisfy({ $0.declarationFragments != nil }) {
+                sortedOverloads = overloadSymbols.sorted(by: { $0.declarationFragments! < $1.declarationFragments! })
+            } else {
+                sortedOverloads = overloadSymbols.sorted(by: { $0.identifier.precise < $1.identifier.precise })
+            }
+            let firstOverload = sortedOverloads.first!
+
+            let overloadGroupIdentifier = firstOverload.identifier.precise + Symbol.overloadGroupIdentifierSuffix
+            var overloadGroupSymbol = firstOverload
+            overloadGroupSymbol.identifier.precise = overloadGroupIdentifier
+            overloadGroupSymbol.isVirtual = true
+
+            let simplifiedDeclaration = overloadGroupSymbol.overloadSubheadingFragments()
+            overloadGroupSymbol.names.navigator = simplifiedDeclaration
+            overloadGroupSymbol.names.subHeading = simplifiedDeclaration
+
+            self.symbols[overloadGroupSymbol.identifier.precise] = overloadGroupSymbol
+
+            // Clone the relationships from the first overload and add them to the overload group
+            for relationship in self.relationships where relationship.source == firstOverload.identifier.precise {
+                var newRelationship = relationship
+                newRelationship.source = overloadGroupIdentifier
+                newRelationships.append(newRelationship)
+            }
+            for relationship in self.relationships where relationship.target == firstOverload.identifier.precise {
+                var newRelationship = relationship
+                newRelationship.target = overloadGroupIdentifier
+                newRelationships.append(newRelationship)
+            }
```

Comment:
Two questions:
- Is the order of the new relationships important? 

Otherwise it would be sufficient to only iterate over all the relationships in the combined graphs _once_.
- Do we want to clone _all_ the relationships? 

If the first overload is an overload, like below
```swift
public class SuperClass {
    public init(_ value: Int) {}
}

public class SomeClass: SuperClass {
    public override init(_ value: Int) {
        super.init(value)
    }
    
    public init(_ value: String) {
        super.init(0)
    }
}
```

or a default implementation of a protocol requirement, like below
```swift
public protocol SomeProtocol {
    func something(_ value: Int)
}

public extension SomeProtocol {
    func something(_ value: Int) {}
}

public class SomeClass: SomeProtocol {
    public func something(_ value: String) {}
}
```

then we could end up with a class having two overloads of the same method or a type having two default implementations of the same protocol requirement. I'm not sure, but that may lead to unexpected behavior in code that processes symbol relationships.

Reply:
Two more questions:

- If the relationship is to a type that's not defined in the local symbol graph, do we still want to clone it?

For example, a method could override a method defined in another module.

- Have we tested overloads for extensions? There is more processing of the relationships for extensions, so we should make sure that that works as expected.

For example, 
```swift
public extension Swift.Int {
    func something(_ value: Int) {}
    func something(_ value: String) {}
}
```

Reply:
Interestingly enough, including default-implementations in the same overload group as the protocol requirement creates a nondeterministic crash in Swift-DocC with some tweaks i have on top of https://github.com/apple/swift-docc/pull/846. I'm wondering if we should just drop default-implementation symbols entirely from this overload-group calculation... 🤔

Reply:
> Is the order of the new relationships important?

It is not; i just felt like the two iterations created easier-to-understand code. I can appreciate not wanting to create excess iteration loops, though, so i can combine them. (I'm already adding enough new loops over the symbols and relationships as it is!)

Reply:
> Have we tested overloads for extensions? There is more processing of the relationships for extensions, so we should make sure that that works as expected.

I just ran a quick test with the Swift-DocC PR and the sample you wrote, and everything worked as expected. As far as SymbolKit is concerned, an extension is just another symbol, and its relationships to the extended type are cloned alongside everything else. Swift-DocC sees the extra symbol in the extension and sorts it properly.

> Do we want to clone _all_ the relationships?

IMO, yes. The purpose of creating the overload group like this is so that it can be used as a stand-in for the first symbol in Swift-DocC for the purposes of curation and linking, but that can otherwise be the parent for all the individual symbols that are overloaded. When you click a link to the overload group symbol, you are taken to a page that should act exactly like you had clicked the "first" symbol, relationships and all.

However, cloning relationships like `defaultImplementationOf` can create other problems, like i mentioned, but that leads me to believe we should be dropping symbols with those relationships from the overload group entirely, rather than dropping those relationships from the cloned symbol.

Reply:
> If the first overload is an overload, like below
> 
> ```swift
> public class SuperClass {
>     public init(_ value: Int) {}
> }
> 
> public class SomeClass: SuperClass {
>     public override init(_ value: Int) {
>         super.init(value)
>     }
>     
>     public init(_ value: String) {
>         super.init(0)
>     }
> }
> ```

This will create a situation where the overload group, whose declaration and docs will match the override due to the way symbols are sorted, will correctly point to the superclass initializer being overridden. The String initializer will not, but will also be part of the overload group. Viewing the page for that initializer will correctly leave out the override relationship.

> or a default implementation of a protocol requirement, like below
> 
> ```swift
> public protocol SomeProtocol {
>     func something(_ value: Int)
> }
> 
> public extension SomeProtocol {
>     func something(_ value: Int) {}
> }
> 
> public class SomeClass: SomeProtocol {
>     public func something(_ value: String) {}
> }
> ```

In this case, if the overload group is created from the protocol requirement, then the `SomeClass` implementation will point to both symbols, yes. However, this is the same default-implementation issue i mentioned earlier: It is a misnomer to consider a protocol requirement and its default implementation to be overloads of each other. In fact, based on some experimentation in Swift-DocC, this can create a crash situation where the overload group is created from the default implementation, where it creates a cyclic ancestor path from the default implementation, to the overload group, to the protocol requirement (where default implementations are automatically curated), then back to the default implementation. This is why i want to drop symbols with `defaultImplementationOf` relationships from overload groups entirely.